### DISCUSSION TITLE
Keycloak 26, user storage provider & XA transactions

### DISCUSSION THREAD

**Discussion by kahlep** [keycloak/keycloak#34182] [Q&A]
Hi,

Keyloak 26 introduces the restriction to have at at most one non-XA datasource.

Our user storage provider configured a separate datasource in a quarkus.properties file and referenced it via the persistence.xml. All this is based on the JPA user storage provider quickstart.
This does not work anymore: it would require XA transactions to be enabled on the datasource but the feature can't be enabled in our Postgres setup according to our DB admin.

I am now looking into using the Keycloak datasource in the plugin and either
- migrate the user store schema to the Keycloak DB host
- set up a foreign data wrapper to provide the data

In this case I would get rid of the quarkus.properties and use the already configured datasource, but how?

We had a similar setup with Keycloak 9 on JBoss with an Oracle DB, where the persistence.xml referenced Keycloak's datasource and the user store was in the same tablespace:
`<jta-data-source>java:jboss/datasources/KeycloakDS</jta-data-source>`
Reverting to this approach, with any datasource name, I get an Exception when building Keycloak:
`java.lang.UnsupportedOperationException: Value found for #getJtaDataSource : not supported yet`

Then I tried to reference the default datasource in different ways (see persistence.xml at the bottom) but all attempts result in this Exception: 
`java.lang.UnsupportedOperationException: The application must supply JDBC connections`

I mainly tried to make the dev mode work with this setup which also requires to autogenerate our schema (previously accomplished through the quarkus.properties: `quarkus.datasource.user_store.jdbc.url=jdbc:h2:mem:user_store;DB_CLOSE_DELAY=-1;INIT=CREATE SCHEMA IF NOT EXISTS USER_STORE`)

The h2 dev mode works with xa transactions enabled and the old setup.
In a deployment with a Postgres DB it worked after removing the quarkus.properties and persistence.xml as the schema is already there, but I'd prefer the dev mode to replicate the production system.

Can someone give me some hints how to make this work?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<persistence version="2.0"
             xmlns="http://java.sun.com/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
             xsi:schemaLocation="
        http://java.sun.com/xml/ns/persistence
        http://java.sun.com/xml/ns/persistence/persistence_2_0.xsd">
    <persistence-unit name="user_store" transaction-type="JTA">
        <class>com.example.UserEntity</class>

        <properties>
            <property name="hibernate.dialect" value="org.hibernate.dialect.H2Dialect" />
            <!-- Sets the name of the datasource to be the same as the datasource name in quarkus.properties-->
            <property name="hibernate.connection.datasource" value="default" />  <!-- this should address the default datasource? -->
            <property name="javax.persistence.transactionType" value="JTA" />
            <property name="hibernate.default_schema" value="user_store" />-->
            <property name="hibernate.hbm2dll.create_namespaces" value="true" />-->
            <property name="hibernate.hbm2ddl.auto" value="update" />
            <property name="hibernate.show_sql" value="true" />
        </properties>
</persistence>
```

**Comment by jordibuj** [comment]
We have a similar case here. We developed a custom user storage provider that retrieves users from an external database. There are several databases, all with the same schema, each one being used for different Keycloak realms.
Each realm configure the custom user storage provider with its assigned database.
So the databases are effectively totally unrelated, they will never participate in the same transaction.
Now we're hitting this restriction where there can only be one non-XA datasource... what can we do in our case?

**Comment by shawkins** [comment]
See https://github.com/keycloak/keycloak/issues/34880#issuecomment-2569310368 - this is addressed in KC 26.1.
